### PR TITLE
Decrease permutation count for fit-symmetry parsing with symmetry groups with more than 2 elements

### PR DESCRIPTION
**Pull request by Stochastic13**
Currently, sampcon converts any symmetry group with more than 2 elements into all combinations of 2-element groups selected from it, resulting in a permutation count of `2^(2Cn)`. In order to remove the redundancy (resulting into `n!` permutation count), I am trying to allow `pyRMSD.RMSDCalculator.OneVsFollowing` to directly parse symmetry groups larger than 2 elements. This only works for fit-symmetry groups since the calculation-symmetry groups are implemented in C++.

### CODE REVIEW

**Comment by Stochastic13** [comment]
I am not sure if there are tests with fit-symmetry groups to check if the modification works for the old 2-element cases. Should I add a test case for >2-element cases?

**Comment by shruthivis** [comment]
Yes it would be preferable to add a test case.

**Comment by Stochastic13** [comment]
> Yes it would be preferable to add a test case.

Alright, I can add a test case to this PR itself. I'll write it based on how the existing tests fare on the new changes once the workflow runs.

File: pyRMSD/RMSDCalculator.py [hunk 1]
```diff
@@ -206,23 +206,18 @@ def oneVsFollowing(self, conformation_number):
             # If we have fitting symmetry groups, we have to try with all possible combinations.
             # Calculation symmetry groups are applied at C level, changing the way RMSD is calculated.
             symm_rmsds = []
-            for permutation in symm_permutations(self.fit_symmetry_groups):
+            original = [numpy.array(i).flatten() for i in self.fit_symmetry_groups]
+            for permutation in symm_permutations_new(self.fit_symmetry_groups):
                 
-                # Copy the coordinates and convert to matrix form for ease of indexing
```
Comment (benmwebb) [inline pyRMSD/RMSDCalculator.py:211]: Don't remove comments unless they really don't make sense any more. Comments are super useful!

File: pyRMSD/symmTools.py [hunk 2]
```diff
@@ -17,6 +18,22 @@ def symm_groups_validation( symm_groups):
                     raise Exception
     except Exception:
         raise ValueError('Symmetry groups are not well defined')
+        
+        
+def symm_groups_validation_new( symm_groups):
```
Comment (benmwebb) [inline pyRMSD/symmTools.py:23]: The name of the function - or at least the docstring - should clarify how it differs from the original `symm_groups_validation`.

File: pyRMSD/symmTools.py [hunk 3]
```diff
@@ -17,6 +18,22 @@ def symm_groups_validation( symm_groups):
...
+    """
+    try:
+        for sg in symm_groups:
+            current_length = None
+            for symm_pair in sg:
+                if current_length is None:
+                    current_length = len(symm_pair)
+                if len(symm_pair) != current_length:
+                    raise Exception
+    except Exception:
```
Comment (benmwebb) [inline pyRMSD/symmTools.py]: This try/except block is unnecessary. Just raise `ValueError` inside the loop. (I know the existing code does it this way.)

File: pyRMSD/symmTools.py [hunk 4]
```diff
@@ -54,6 +71,46 @@ def symm_permutations(groups):
...
+
+def symm_permutations_new(groups):
+    """
+    Generator that produces the possible permutations of the symmetry groups.
+    Suposing we have this symmetries definition:
+        [
+            [ [1,2,0] ], 
+            [ [3,4],[5,6] ]
+        ]
+    Meaning that coordinates 1 and 2 and 3 are equivalent, and that 3 and 4, as well as 5 and 6 are equivalen and must be changed
```
Comment (benmwebb) [inline pyRMSD/symmTools.py]: Don't you mean "coordinates 1 and 2 and **0** are equivalent" ?

File: pyRMSD/symmTools.py [hunk 5]
```diff
@@ -54,6 +71,46 @@ def symm_permutations(groups):
...
+    if len(groups) > 0:
+        head = groups[0]
+        for perm in permutations(range(len(head[0]))):
+            swapped_head = []
+            for item in head:
+                swapped_head.append([item[i] for i in perm])
+            for tail_permutation in symm_permutations_new(groups[1:]):
+                yield [swapped_head] + tail_permutation
+    else:
+        yield []
```
Comment (benmwebb) [inline pyRMSD/symmTools.py:110]: This seems wrong to me (if there are no groups, there should be no permutations, not one permutation that is empty). But the original code does it this way too so I guess that's OK.

File: pyRMSD/test/testRMSDCalculators.py [hunk 6]
```diff
@@ -249,6 +249,32 @@ def test_fitting_symmetry(self):
                                                                              ])
         rmsds = calculator.oneVsFollowing(0)
         numpy.testing.assert_almost_equal(expected_rmsds, rmsds, 12)
+        
+        
+    def test_fitting_symmetry_new(self):
```
Comment (benmwebb) [inline pyRMSD/test/testRMSDCalculators.py:254]: Test should have a docstring to describe what it's trying to test.

File: pyRMSD/test/testRMSDCalculators.py [hunk 7]
```diff
@@ -249,6 +249,32 @@ def test_fitting_symmetry(self):
...
+        protein_skeleton = numpy.array([[0, 0, 0],[0, 1, 0],[0, 0, 1],[1, 0, 0],[1, 1, 1]])
+        thetas = [0, numpy.pi / 6, numpy.pi / 2, numpy.pi]
+        proteins = []
+        # axis is (1, 0, 0)
+        for i in range(4):
+            t = thetas[i]
+            rot_matrix = numpy.array([[1, 0, 0], [0, numpy.cos(t), -numpy.sin(t)], [0, numpy.sin(t), numpy.cos(t)]])
+            proteins.append((rot_matrix @ protein_skeleton.T).T + i * 3)
```
Comment (benmwebb) [inline pyRMSD/test/testRMSDCalculators.py]: Use explicit matrix multiply rather than `@` operator (which doesn't exist in Python 2).

File: pyRMSD/test/testSymmTools.py [hunk 8]
```diff
@@ -37,9 +39,13 @@ def test_validation(self):
...
             self.fail("Value exception has been raised.")
     
-        symm_group_3 = [ [2,1,3] ] # A symm group must have elements of len 2
-         
+        # Old validation check fails
         self.assertRaises(ValueError, symm_groups_validation, [symm_group_3])
+        # New validation check passes
+        try:
+            symm_groups_validation_new([symm_group_3])
+        except ValueError:
```
Comment (benmwebb) [inline pyRMSD/test/testSymmTools.py]: This is unnecessary. If the function raises `ValueError`, it will fail the test already.

File: pyRMSD/test/testSymmTools.py [hunk 9]
```diff
@@ -58,6 +64,35 @@ def test_symm_group_permutations(self):
         for i, perm in enumerate(symm_permutations(groups)):
             self.assertSequenceEqual(expected_permutations[i], perm)
             
+    def test_symm_group_permutations_new(self):
```
Comment (benmwebb) [inline pyRMSD/test/testSymmTools.py:64]: No docstring. Should also test the return value when given an empty list of groups.

**Review by benmwebb — CHANGES_REQUESTED** [review]
Looks mostly OK to me, once the tests are passing.

File: pyRMSD/RMSDCalculator.py (hunk 1 above)
Comment (Stochastic13) [reply pyRMSD/RMSDCalculator.py:211]: Added back

File: pyRMSD/symmTools.py (hunk 2 above)
Comment (Stochastic13) [reply pyRMSD/symmTools.py:23]: Added in the docstring

File: pyRMSD/symmTools.py (hunk 3 above)
Comment (Stochastic13) [reply pyRMSD/symmTools.py]: Removed the try-except block.

File: pyRMSD/symmTools.py (hunk 4 above)
Comment (Stochastic13) [reply pyRMSD/symmTools.py]: Fixed in the new PR

File: pyRMSD/test/testRMSDCalculators.py (hunk 6 above)
Comment (Stochastic13) [reply pyRMSD/test/testRMSDCalculators.py:254]: Added docstrings in the two tests

File: pyRMSD/test/testRMSDCalculators.py (hunk 7 above)
Comment (Stochastic13) [reply pyRMSD/test/testRMSDCalculators.py]: Fixed in the new PR

File: pyRMSD/test/testSymmTools.py (hunk 9 above)
Comment (Stochastic13) [reply pyRMSD/test/testSymmTools.py:64]: Added

File: pyRMSD/test/testSymmTools.py (hunk 8 above)
Comment (Stochastic13) [reply pyRMSD/test/testSymmTools.py]: Removed the try-except block

**Review by benmwebb — APPROVED** [review]
